src/Bicep.Core/TypeSystem/DeclaredTypeManager.cs

```diff
@@ -759,7 +759,146 @@ private ITypeReference GetUnionTypeType(UnionTypeSyntax syntax)
+                && TryGetSystemDecorator(decorableSyntax, LanguageConstants.TypeDiscriminatorDecoratorName) is { } discriminatorDecorator)
+            {
+                return FinalizeDiscriminatedObjectType(unionMembers.ToArray(), discriminatorDecorator);
+            }
+
+            return TypeHelper.CreateTypeUnion(unionMembers.Select(t => t.Item2));
+        }
+
+        private DecorableSyntax? TryResolveUnionImmediateDecorableSyntax(SyntaxBase? syntaxBase) =>
+            syntaxBase switch
+            {
+                TypeDeclarationSyntax or ObjectTypePropertySyntax => syntaxBase as DecorableSyntax,
+                ParenthesizedExpressionSyntax or UnionTypeSyntax or UnionTypeMemberSyntax =>
+                    TryResolveUnionImmediateDecorableSyntax(binder.GetParent(syntaxBase)),
+                _ => null
+            };
+
+        private TypeSymbol FinalizeDiscriminatedObjectType(
+            (UnionTypeMemberSyntax syntax, ITypeReference type)[] unionMembers, DecoratorSyntax discriminatorDecorator)
+        {
+            var discriminatorPropertyExpr = discriminatorDecorator.Arguments.FirstOrDefault()?.Expression as StringSyntax;
+            var discriminatorPropertyName = discriminatorPropertyExpr?.TryGetLiteralValue();
+
+            if (discriminatorPropertyName == null)
+            {
+                return ErrorType.Empty(); // the decorator validator handles this case
+            }
+
+            List<ErrorDiagnostic>? errorDiagnostics = null;
+
+            void AddErrorDiagnostic(ErrorDiagnostic errorDiagnostic)
+            {
+                if (errorDiagnostics == null)
+                {
+                    errorDiagnostics = new List<ErrorDiagnostic>();
+                }
+
+                errorDiagnostics.Add(errorDiagnostic);
+            }
+
+            var memberDiscriminatorValues = new HashSet<string>();
+            var expandedMemberTypes = new List<ObjectType>();
+
+            foreach (var (memberSyntax, memberType) in unionMembers)
+            {
+                var memberTypeEvaluated = memberType.Type;
+
+                if (memberTypeEvaluated is ObjectType objectType)
+                {
+                    // validate the member has the discriminator property defined
+                    if (!objectType.Properties.TryGetValue(discriminatorPropertyName, out var discriminatorTypeProperty))
+                    {
+                        AddErrorDiagnostic(DiagnosticBuilder.ForPosition(memberSyntax)
+                            .DiscriminatorPropertyMustBeRequiredStringLiteral(discriminatorPropertyName));
+
+                        continue;
+                    }
+
+                    // validate the discriminator property is of the right type
+                    if (discriminatorTypeProperty.TypeReference.Type.TypeKind != TypeKind.StringLiteral
+                        || !discriminatorTypeProperty.Flags.HasFlag(TypePropertyFlags.Required))
```

Comment:
```suggestion
                    if (discriminatorTypeProperty.TypeReference.Type.TypeKind is not StringLiteralType discriminatorType
                        || !discriminatorTypeProperty.Flags.HasFlag(TypePropertyFlags.Required))
```

That way you can use `discriminatorType.RawStringValue` on line 848 without needing to redo this check on lines 845-846

Reply:
[dd5cea7](https://github.com/Azure/bicep/pull/11222/commits/dd5cea7ec62d7b867977e6ea63253b13da2e90d4)